utilities/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+// #[macro_use]
+
+use std::f64;
+use std::fs::File;
+use std::io::{self, BufRead};
+use std::iter::Iterator;
+use std::str;
+use std::vec::Vec;
+
+
+// todo: remove the *.dat files from residing directly in this repo, and instead pull them externally?
+// todo: switch from truncated *.dat files to full *.dat files (after getting them out of this repo)
+
+pub const NON_NUMS_NONE: &[usize; 0] = &[];
+
+
+// Return the absolute and relative difference between two values, in that order.
+// If both values are nan, consider the difference to be 0.
+pub fn calc_delta(x: f64, y: f64) -> (f64, f64) {
+  if x.is_nan() && y.is_nan() {
+    return (0f64, 0f64);
```

Comment:
As with most things NaN, I have no idea if this is reasonable.

Glad to see the introduction of relative eq check. I mostly use it via the `approx` crate, but I don't think I knew about it when I started working on this crate.

Reply:
We definitely want to consider a pair of NaNs to be equal in most case, even though f64::NAN==f64::NAN returns false, and I'm pretty sure f64::NAN-f64::NAN returns f64::NAN, because we want to consider the C++ and rust results to match if they return the same value, even when that value is NaN.

Similarly, we want the option to consider 0.0 and -0.0 to not match, even though I'm pretty sure the equality operator returns true, and their difference is 0. The same is true for NAN and -NAN.

I'll make a note to explore the approx crate. It looks pretty promising, but I'm not yet clear whether it can be easily coaxed into the desired handling for all the cases above. If it can, it would be great to retire my custom comparison logic, but I probably won't explore this in my current pass.

Reply:
Thanks for elaborating. I think it's totally fine to keep this snippet here - no need to spend time digging into `approx`.